midend/local_copyprop.cpp

```diff
@@ -55,58 +59,79 @@ void P4::DoLocalCopyPropagation::flow_merge(Visitor &a_) {
 }
 
 void P4::DoLocalCopyPropagation::dropValuesUsing(cstring name) {
+    LOG6("dropValuesUsing(" << name << ")");
     for (auto &var : available) {
+        LOG7("  checking " << var.first << " = " << var.second.val);
         if (var.first == name) {
             LOG4("   dropping " << name << " as it is being assigned to");
             var.second.val = nullptr;
         } else if (var.second.val && exprUses(var.second.val, name)) {
-            LOG4("   dropping " << var.first << " as it is use" << name);
+            LOG4("   dropping " << var.first << " as it uses " << name);
             var.second.val = nullptr; } }
 }
 
 const IR::Node *P4::DoLocalCopyPropagation::postorder(IR::Declaration_Variable *var) {
-    LOG1("Visiting " << getOriginal());
     if (!in_action) return var;
+    LOG1("Visiting " << getOriginal());
     if (available.count(var->name))
         BUG("duplicate var declaration for %s", var->name);
     auto &local = available[var->name];
     local.local = true;
     if (var->initializer) {
         if (!hasSideEffects(var->initializer)) {
-            LOG3("  saving init value for " << var->name);
+            LOG3("  saving init value for " << var->name << ": " << var->initializer);
             local.val = var->initializer;
         } else {
             local.live = true; } }
     return var;
 }
 
 const IR::Expression *P4::DoLocalCopyPropagation::postorder(IR::PathExpression *path) {
+    if (!in_action) return path;
+    if (isWrite()) {
+        dropValuesUsing(path->path->name);
+        return path; }
     if (auto var = ::getref(available, path->path->name)) {
-        if (isWrite()) {
-            return path;
-        } else if (var->val) {
-            LOG3("  propagating value for " << path->path->name);
+        if (var->val) {
+            LOG3("  propagating value for " << path->path->name << ": " << var->val);
             return var->val;
         } else {
             LOG4("  using " << path->path->name << " with no propagated value");
             var->live = true; } }
     return path;
 }
 
+IR::AssignmentStatement *P4::DoLocalCopyPropagation::preorder(IR::AssignmentStatement *as) {
+    if (!in_action) return as;
+    // visit the source subtree first, before the destination subtree
+    // make sure child indexes are set properly so we can detect writes
+    visit(as->right, "right", 1);
+    visit(as->left, "left", 0);
```

Comment:
There is one subtle case, which I can't figure out whether it's handled properly, because the mixin code is in a separate place. Not everything on the LHS of an assignment is in a write context. The only exception currently is the array index:

a[x].b = c;

In this case x is just read, not written.